src/ODEs/ODETools/EXRungeKutta.jl

```diff
@@ -0,0 +1,148 @@
+"""
+Explicit Runge-Kutta ODE solver
+"""
+struct EXRungeKutta <: ODESolver
+  nls::NonlinearSolver
+  dt::Float64
+  tableau::EXButcherTableau
+  function EXRungeKutta(nls::NonlinearSolver, dt, type::Symbol)
+    bt = EXButcherTableau(type)
+    new(nls, dt, bt)
+  end
+end
+
+"""
+solve_step!(uf,odesol,op,u0,t0,cache)
+"""
+function solve_step!(uf::AbstractVector,
+  solver::EXRungeKutta,
+  op::ODEOperator,
+  u0::AbstractVector,
+  t0::Real,
+  cache)
+
+  # Unpack variables
+  dt = solver.dt
+  s = solver.tableau.s
+  a = solver.tableau.a
+  b = solver.tableau.b
+  c = solver.tableau.c
+  d = solver.tableau.d
+
+  # Create cache if not there
+  if cache === nothing
+    ode_cache = allocate_cache(op)
+    vi = similar(u0)
+    ki = [similar(u0) for i in 1:s]
+    M = allocate_jacobian(op,t0,uf,ode_cache)
+    get_mass_matrix!(M,op,t0,uf,ode_cache)
+    nl_cache = nothing
+  else
+    ode_cache, vi, ki, M, nl_cache = cache
+  end
+
+  nlop = EXRungeKuttaStageNonlinearOperator(op,t0,dt,u0,ode_cache,vi,ki,0,a,M)
+
+  for i in 1:s
+
+    # solve at stage i
+    ti = t0 + c[i]*dt
+    ode_cache = update_cache!(ode_cache,op,ti)
+    update!(nlop,ti,ki[i],i)
+    nl_cache = solve!(uf,solver.nls,nlop,nl_cache)
+
+    update!(nlop,ti,uf,i)
+
+  end
+
+  # update final solution
+  tf = t0 + dt
+
+  @. uf = u0
+  for i in 1:s
+  @. uf = uf + dt*b[i]*nlop.ki[i]
+  end
+
+  cache = (ode_cache, vi, ki, M, nl_cache)
+
+  return (uf,tf,cache)
+
+
+end
+
+
+
+
+mutable struct EXRungeKuttaStageNonlinearOperator <: RungeKuttaNonlinearOperator
+  odeop::ODEOperator
+  ti::Float64
+  dt::Float64
+  u0::AbstractVector
+  ode_cache
+  vi::AbstractVector
+  ki::AbstractVector
+  i::Int
+  a::Matrix
+  M::AbstractMatrix
+end
+
+
+"""
+ODE:    A(t,u,∂u  = M ∂u/∂t + K(t,u) = 0 -> solve for u
+EX-RK:  A(t,u,ki) = M ki    + K(ti,u0 + dt ∑_{j<i} a_ij * kj) = 0 -> solve for ki
+                  = M ki    + K(ti,ui) = 0
+
+For forward euler,          i = 1     -> ui = u0
```

Comment:
don't you think this is just a particular case of the general case in which there is only one stage, I think you can accommodate it in the same implementation / def'on

Reply:
updated this comment to be more general in clean version